### ISSUE TITLE
Session name unique within scenario scope

### ISSUE THREAD

**Issue by sponte** [Stub-O-Matic/stubo-app#93]
Thanks for implementing #52. This is working as expected. I am able to rename scenarios (tested on 0.6.4). 

However, I have discovered that after renaming the scenario, when I try to reimport stubs again from the same source, it fails due to duplicate session id. Can you make session id unique within the context of scenario? There was a ticket #39 that was supposed to add parameters to the exported commands however I tried on 0.6.4 and parameters are not there. 

Either solution will work for us, with later requiring us to regenerate all stubs that we have.

**Comment by rusenask** [comment]
Global unique scenario thing puzzled me from the beginning, it really makes little sense and gives you only one benefit - skip scenario name when you want to begin session although there are quite a lot of disadvantages from it, such as it stubo has to look for scenario itself and it's harder to manage sessions from the client. This would be a major change in current design so more discussions regarding session and it's status should take place. Thanks for bringing this up!

**Comment by rusenask** [comment]
@sponte - I could change the API and cache logic (which stores session-scenario mapping) although prior to that we should consider what you would have to do on your side as well. Are you aware that you would need to change these calls?
http://stubo-app.readthedocs.org/en/latest/pages/api.html#get-response
http://stubo-app.readthedocs.org/en/latest/pages/api.html#end-session
I believe your integrators would require modifications.

**Comment by sponte** [comment]
@rusenask - not sure what the best approach here is. This indeed is a massive change that would require lots of maintenance work for existing users. Having said that, we should go with a right solution and not the one that yields less maintenance work.

Alternatively - maybe we can do something inside import stub call. If stubo can generate dynamic session id for import stub based on scenario name and some unique id, this would resolve the original problem I was having and the rest should still work as previously. What do you think?

**Comment by rusenask** [comment]
@sponte - in the long run the right solution would be to make session unique only to scenario, John D. agrees to that. So it will probably be added to our roadmap. 

For unique sessions - I guess we could just use Python's UUID module to generate unique IDs if you just need a unique name. How do you pass session names down to the integrators? Also, rename scenario function could also rename sessions..

**Comment by sponte** [comment]
short term renaming both parameters would work. Our temporary workflow would be as follows:

1. Import stubs
2. Rename scenario name and session id
3. Run our tests
4. Delete scenario and all sessions within it

Integrators have library which has start new stubo session functionality which takes session id.

**Comment by rusenask** [comment]
what about supplying both scenario name and session id in the same field?
something like this:
stubo/api/get/response?session=scenario_10:first_1

This way you wouldn't have to update your integrators to take scenario parameter. Also, I will then expand rename scenario function to rename sessions and stubs (since they store session name) as well. 
Having this format will make session name unique only to scenarios (since sessions will include scenario names inside them). Sessions, that are created without scenario namespace will be "global" sessions and I will add additional check to dismiss begin session request which try to create a scenario and session with different scenario names:
stubo/api/begin/session?scenario=first&session=second:first_1&mode=record
Example above would be treated as bad request. 

What do you think about it?

**Comment by sponte** [comment]
Sorry, not ignoring, just consulting with others

**Comment by sponte** [comment]
Had a chat with my colleague and we think adding ability to change session name would be the simplest and easiest solution for now. Hacking around session name seems a bit complicated and could have some side effects.

**Comment by rusenask** [comment]
By "hacking around with session name" you mean creating additional API call to change session name? It's not really a simplest and easiest solution :) If sessions were stored in database - then it probably wouldn't be very hard, however they are stored in cache now and in some interesting structures. I will discuss this further with other stubo members but we will probably go with new format for defining scenario sessions "scenario_name:session_name". It won't require any changes on your integrator side.

**Comment by rusenask** [comment]
It appears that this happened before:
https://github.com/Stub-O-Matic/stubo-app/blob/master/stubo/service/handlers_mt.py#L269-L305
Here we can see remains of "legacy" format where both scenario and session names were required in headers. I wonder what happened and why they changed the format. Implementation is awful since you can just use  self.request.headers.get('session', None) instead of cycles. However this code would fail now anyway since documentation doesn't mention that you need to pass scenario in headers.

**Comment by johnecobo** [comment]
The session / scenario design was based on the following situation:
We have two test teams, one in London and one in Seattle, Both have test suites which test passenger checkin, call this this test scenario = checkin_Bob." The goal is for the London team and the Seattle team to be able to execute scenario checkin_Bob whenever they want without needing to confer and plan with the other team to avoid clashes.

Another challenge is that if the Seattle team start the test first, we don't want the London team to accidentally end Seattle's checkin_Bob session.

The "session" allows multiple concurrent runs of the same scenario. The problem has come because of the "hard coding" of the session names in command files, but I thought there was a recent change to make them dynamically entered as part of calling the command file.

Once a session is running, it can be ended with just the session name (thus must be globally unique). The idea was that the session names can be long-ish and readable to humans. Session names are more of a short id. 

If Stubo were to generate the session name, how would the client (test tool) know that session name to end it? If the scenario and session were combined into one field, what have you gained?

**Comment by rusenask** [comment]
If session name was generated by stubo - it would return this ID back to the client so the test tool would grab this variable and use it later to end the session. 

Regarding scenario_name:session_name - it would  make sessions not global but unique to each scenario. In cache they would be stored as hostname:scenario:session so their test workflow could be like this:
* Import scenario with stubs
* Change scenario name
* Do the tests
* Delete scenario with stubs

**Comment by JohnFDavenport** [comment]
@johnecobo - Is the scenario you mention one that was likely to occur in Stubo Classic, where there was typically a limited set of Stubo instances distinguished by a few IP addresses, but unlikely in the current Stubo as we now recommend users manage separation using potentially infinite numbers of unique URLs (or hosts)? Granted it might happen if teams were piling on to the same system to record and playback tests, but we know from experience that is an undisciplined approach and other bad things happen that are not solvable by anything Stubo can do. Users have to be aware now that need to manage access within the host, but the host is effectively fully virtualised. Just trying to understand how this might occur as I think Seattle and London would be using separate URLs/hosts.

**Comment by johnecobo** [comment]
You are correct that the clash of scenarios is less likely to happen now. Stubo can visualize by URL IF:
* different test systems are in use - pointing to different Stubo URLs, or
* testers and tools are able to pass in the Stubo URL as an argument to the test and so even if sharing a single integration test environment can use different Stubo URLs.

**Comment by johnecobo** [comment]
A better solution is to use wildcarded Stubo URLs to achieve virtual, private Stubos for each situation where they are needed. Companies should try to do so.
